src/handlers/notify_zulip.rs

```diff
@@ -1,96 +1,171 @@
+    global_config: &NotifyZulipConfig,
+) -> Vec<NotifyZulipInput> {
+    event
+        .issue
+        .labels
+        .iter()
+        .cloned()
+        .filter_map(|label| {
+            global_config
+                .labels
+                .get(&label.name)
+                .map(|config| (label, config))
+        })
+        .flat_map(|(label, config)| {
+            if !has_all_required_labels(&event.issue, config) {
+                // Issue misses a required label, ignore this event
+                return None;
+            }
+
+            match event.action {
+                IssuesAction::Closed if config.message_on_close.is_some() => {
+                    Some(NotifyZulipInput {
+                        notification_type: NotificationType::Closed,
+                        label,
+                    })
+                }
+                IssuesAction::Reopened if config.message_on_reopen.is_some() => {
+                    Some(NotifyZulipInput {
+                        notification_type: NotificationType::Reopened,
+                        label,
+                    })
                 }
+                _ => None,
             }
+        })
+        .collect()
+}
 
-            if event.action == IssuesAction::Labeled && config.message_on_add.is_some() {
-                return Ok(Some(NotifyZulipInput {
-                    notification_type: NotificationType::Labeled,
-                }));
-            } else if config.message_on_remove.is_some() {
-                return Ok(Some(NotifyZulipInput {
-                    notification_type: NotificationType::Unlabeled,
-                }));
+fn has_all_required_labels(issue: &Issue, config: &NotifyZulipLabelConfig) -> bool {
+    for req_label in &config.required_labels {
+        let pattern = match glob::Pattern::new(req_label) {
+            Ok(pattern) => pattern,
+            Err(err) => {
+                log::error!("Invalid glob pattern: {}", err);
+                continue;
             }
+        };
+        if !issue.labels().iter().any(|l| pattern.matches(&l.name)) {
+            return false;
         }
     }
-    Ok(None)
+
+    true
 }
 
 pub(super) async fn handle_input<'a>(
     ctx: &Context,
     config: &NotifyZulipConfig,
     event: &IssuesEvent,
-    input: NotifyZulipInput,
+    inputs: Vec<NotifyZulipInput>,
 ) -> anyhow::Result<()> {
-    let config = config
-        .labels
-        .get(&event.label.as_ref().unwrap().name)
-        .unwrap();
-
-    let mut topic = config.topic.clone();
-    topic = topic.replace("{number}", &event.issue.number.to_string());
-    topic = topic.replace("{title}", &event.issue.title);
-    // Truncate to 60 chars (a Zulip limitation)
-    let mut chars = topic.char_indices().skip(59);
-    if let (Some((len, _)), Some(_)) = (chars.next(), chars.next()) {
-        topic.truncate(len);
-        topic.push('…');
-    }
+    for input in inputs {
+        let config = config.labels.get(&input.label.name).unwrap();
```

Comment:
Nit: I think this is equivalent to
```suggestion
        let config = config.labels[&input.label.name];
```